### PR TITLE
Use Packet struct to define a test packet

### PR DESCRIPTION
**Pull request by mazdakn**
## Description
This PR refactors the test packet creation to be extensible. This is needed so new protocols or headers like IPv6 extensions can be added easily.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

File: felix/bpf/ut/bpf_prog_test.go [hunk 1]
```diff
@@ -864,189 +893,164 @@ var udpDefault = &layers.UDP{
...
+	}
 
-		} else {
-			ipv4.Length = uint16(5*4 + 8 + len(payload))
-			_ = tcp.SetNetworkLayerForChecksum(ipv4)
+	if pkt.isIPv6 {
+		pkt.handleIPv6()
+	} else {
+		pkt.handleIPv4()
+	}
```
Comment (tomastigera) [inline felix/bpf/ut/bpf_prog_test.go:1029]: you could do `pkt.handleL3` in a similar vein as `handleL4`.

File: felix/bpf/ut/bpf_prog_test.go [hunk 2]
```diff
@@ -864,189 +893,164 @@ var udpDefault = &layers.UDP{
...
-		if ipv6Alt != nil {
-			ipv6 = ipv6Alt
+	pkt.length += 8
+	return nil
+}
+
+func (pkt *Packet) setChecksum() {
+	switch pkt.protocol {
+	case layers.IPProtocolUDP:
+		if pkt.isIPv6 {
```
Comment (tomastigera) [inline felix/bpf/ut/bpf_prog_test.go:967]: Do you need the `isIPv6` You could have a method of  Packet.L3() that would return the L3 layer which ever is set or panic if both are set.

File: felix/bpf/ut/bpf_prog_test.go [hunk 3]
```diff
@@ -864,189 +893,164 @@ var udpDefault = &layers.UDP{
...
-			udp.Length = uint16(8 + len(payload))
-
-			pkt := gopacket.NewSerializeBuffer()
-			err := gopacket.SerializeLayers(pkt, gopacket.SerializeOptions{ComputeChecksums: true},
-				eth, ipv6, udp, gopacket.Payload(payload))
-			return eth, nil, ipv6, udp, payload, pkt.Bytes(), err
+func (pkt *Packet) handleEthernet() {
+	if pkt.eth == nil {
+		if pkt.isIPv6 {
+			pkt.eth = ethDefaultWithIPv6
```
Comment (tomastigera) [inline felix/bpf/ut/bpf_prog_test.go:1010]: similarly you could have `Packet.L2Proto()`  and set the appropriate proto in the eth header, the MACs are the same. And we could also check that the supplied eth proto isnt nonsense based on the L3, perhaps, but it is test ...

**Review by tomastigera — APPROVED** [review]
Great clean up! Thanks!

File: felix/bpf/ut/bpf_prog_test.go (hunk 2 above)
Comment (mazdakn) [reply felix/bpf/ut/bpf_prog_test.go:967]: removed it.

**Pull request merged**